File: app/core/tracking.py
from typing import Callable, Optional

import cv2
import numpy as np

from app.core.image_sequence import ImageSequence
from app.core.result import TrackerResult

# Flat LK parameters (assembled into cv2 kwargs by _cv_lk_kwargs).
DEFAULT_LK = dict(
    win_size=21,
    max_level=3,
    max_iter=30,
    epsilon=0.01,
    flags=0,
    min_eig_threshold=1e-4,
)

ProgressCb = Callable[[int, int], bool]  # (done, total) -> cancel?


def _cv_lk_kwargs(p: dict) -> dict:
    required = set(DEFAULT_LK)
    missing = sorted(required - set(p))
    if missing:
        raise ValueError(f"Missing LK parameter(s): {', '.join(missing)}")
    try:
        win_size = int(p["win_size"])
        max_level = int(p["max_level"])
        max_iter = int(p["max_iter"])
        epsilon = float(p["epsilon"])
        flags = int(p["flags"])
        min_eig = float(p["min_eig_threshold"])
    except (TypeError, ValueError) as exc:
        raise ValueError("LK parameters must be numeric") from exc
    if win_size < 3 or max_level < 0 or max_iter < 1 or epsilon <= 0 or min_eig < 0:
        raise ValueError("LK parameters are outside their valid ranges")
    if flags not in (0, cv2.OPTFLOW_LK_GET_MIN_EIGENVALS):
        raise ValueError("LK flags must select standard error (0) or minimum-eigenvalue error (8)")
    if not np.isfinite(epsilon) or not np.isfinite(min_eig):
        raise ValueError("LK parameters must be finite")
    return dict(
        winSize=(win_size, win_size),
        maxLevel=max_level,
        criteria=(
            cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT,
            max_iter,
            epsilon,
        ),
        flags=flags,
        minEigThreshold=min_eig,
    )


def _track_pass(
    get_gray,
    frame_order,
    seed_pts,
    lk_kwargs,
    progress,
    total,
    done0,
    seed_valid=None,
):
    """Run LK along `frame_order` (list of global indices). Results are stored at the
    *position within frame_order* (index 0 = the seed frame)."""
    n = len(frame_order)
    p = len(seed_pts)
    coords = np.zeros((n, p, 2), dtype=np.float32)
    status = np.zeros((n, p), dtype=np.uint8)
    err = np.zeros((n, p), dtype=np.float32)

    coords[0] = seed_pts
    alive = (
        np.ones(p, dtype=bool)
        if seed_valid is None
        else np.asarray(seed_valid, dtype=bool).reshape(p).copy()
    )
    status[0] = alive.astype(np.uint8)
    err[0, ~alive] = np.inf

    prev_gray = get_gray(frame_order[0])
    prev_pts = seed_pts.reshape(-1, 1, 2).astype(np.float32)
    done = done0
    for i in range(1, n):
        cur_gray = get_gray(frame_order[i])
        nxt, st, er = cv2.calcOpticalFlowPyrLK(
            prev_gray, cur_gray, prev_pts, None, **lk_kwargs
        )
        # OpenCV's status describes only this one frame-to-frame transition. A point that has
        # already been lost must never become valid again: its returned location is undefined and
        # feeding that location into the next transition can create a plausible-looking "revived"
        # track. Keep validity cumulative and freeze dead points at their last valid coordinate.
        if nxt is None or st is None:
            next_pts = coords[i - 1].copy()
            step_valid = np.zeros(p, dtype=bool)
            step_err = np.full(p, np.inf, dtype=np.float32)
        else:
            next_pts = np.asarray(nxt, dtype=np.float32).reshape(-1, 2)
            step_valid = np.asarray(st).reshape(-1).astype(bool)
            if next_pts.shape != (p, 2) or step_valid.shape != (p,):
                raise ValueError("OpenCV returned an unexpected optical-flow result shape")
            step_valid &= np.isfinite(next_pts).all(axis=1)
            step_err = (
                np.asarray(er, dtype=np.float32).reshape(-1)
                if er is not None
                else np.full(p, np.inf, dtype=np.float32)
            )
            if step_err.shape != (p,):
                raise ValueError("OpenCV returned an unexpected optical-flow error shape")
            step_err[~np.isfinite(step_err)] = np.inf

        alive &= step_valid
        next_pts[~alive] = coords[i - 1, ~alive]
        step_err[~alive] = np.inf
        coords[i] = next_pts
        status[i] = alive.astype(np.uint8)
        err[i] = step_err
        prev_gray = cur_gray
        prev_pts = next_pts.reshape(-1, 1, 2)
        done += 1
        if progress is not None and progress(done, total):
            return None
    return coords, status, err, done


def _compute_fb_errors(coords_fw, coords_bw, status_fw, status_bw):
    diff = np.linalg.norm(coords_fw - coords_bw, axis=2)  # (N, P)
    valid = (status_fw == 1) & (status_bw == 1)  # (N, P)
    if diff.shape[0] == 1:
        zeros = np.zeros(diff.shape[1], dtype=np.float32)
        return zeros, zeros.copy()
    # The last frame is the backward seed, so its discrepancy is identically zero and carries no
    # round-trip information. Including it would systematically bias every FB mean downward.
    valid[-1] = False
    p = diff.shape[1]
    fb_mean = np.full(p, np.inf, dtype=np.float32)
    fb_max = np.full(p, np.inf, dtype=np.float32)
    for j in range(p):
        col = diff[valid[:, j], j]
        if col.size:
            fb_mean[j] = col.mean()
            fb_max[j] = col.max()
    return fb_mean, fb_max


def track(
    sequence: ImageSequence,
    reference_index: int,
    last_index: int,
    seed_pts: np.ndarray,
    lk_params: dict,
    progress_cb: Optional[ProgressCb] = None,
) -> Optional[TrackerResult]:
    """Track seed_pts forward (reference->last) then backward (last->reference, seeded from
    the forward result's last-frame positions). Returns None if cancelled via progress_cb.

    coords_bw is reindexed so coords_bw[t] aligns with coords_fw[t] (cut 0 = reference).
    """
    seed_pts = np.asarray(seed_pts, dtype=np.float32).reshape(-1, 2)
    if not 0 <= reference_index <= last_index < len(sequence):
        raise ValueError("Tracking range is outside the loaded image sequence")
    if seed_pts.shape[0] == 0 or not np.isfinite(seed_pts).all():
        raise ValueError("Tracking requires at least one finite seed point")
    h, w = sequence.load_gray(reference_index).shape[:2]
    if np.any(seed_pts < 0) or np.any(seed_pts[:, 0] >= w) or np.any(seed_pts[:, 1] >= h):
        raise ValueError("Tracking seeds must be inside the reference image")
    n = last_index - reference_index + 1
    lk_kwargs = _cv_lk_kwargs(lk_params)
    total = 2 * (n - 1)

    def get_gray(global_index):
        return sequence.load_gray(global_index)

    forward_order = list(range(reference_index, last_index + 1))
    fw = _track_pass(get_gray, forward_order, seed_pts, lk_kwargs, progress_cb, total, 0)
    if fw is None:
        return None
    coords_fw, status_fw, err_fw, done = fw

    # Backward pass: seed from forward's last-frame positions, walk last -> reference.
    backward_order = list(range(last_index, reference_index - 1, -1))
    bw_seed = coords_fw[n - 1]
    # A point that did not survive the complete forward pass has no trustworthy last-frame seed.
    # Keep it invalid throughout the backward pass so FB metrics cannot be rescued by garbage.
    bw_seed_valid = status_fw[n - 1].astype(bool)
    bw = _track_pass(
        get_gray,
        backward_order,
        bw_seed,
        lk_kwargs,
        progress_cb,
        total,
        done,
        seed_valid=bw_seed_valid,
    )
    if bw is None:
        return None
    coords_bw_rev, status_bw_rev, err_bw_rev, _ = bw

    # Reindex backward arrays from pass-order (last..reference) to cut-order (reference..last).
    coords_bw = coords_bw_rev[::-1].copy()
    status_bw = status_bw_rev[::-1].copy()
    err_bw = err_bw_rev[::-1].copy()

    fb_mean, fb_max = _compute_fb_errors(coords_fw, coords_bw, status_fw, status_bw)

    return TrackerResult(
        reference_index=reference_index,
        last_index=last_index,
        coords_fw=coords_fw,
        status_fw=status_fw,
        err_fw=err_fw,
        coords_bw=coords_bw,
        status_bw=status_bw,
        err_bw=err_bw,
        fb_mean_error=fb_mean,
        fb_max_error=fb_max,
        win_size=int(lk_params["win_size"]),
        error_kind="min_eigenvalue" if int(lk_params.get("flags", 0)) & 8 else "photometric",
        tracking_params=dict(lk_params),
    )

File: rust/crates/core/src/tracking.rs
//! Pyramidal Lucas-Kanade tracking: forward (reference->last) then backward
//! (last->reference, seeded from the forward pass's last frame), plus the per-point
//! forward-backward error. Ports `app/core/tracking.py`.

use crate::image_sequence::ImageSequence;
use crate::result::{LkParams, TrackerResult};
use ndarray::{s, Array1, Array2, Array3};
use opencv::core::{Point2f, Size, TermCriteria, TermCriteria_Type, Vector};
use opencv::video;

/// `(done, total) -> cancel?`
pub type ProgressCb<'a> = &'a mut dyn FnMut(usize, usize) -> bool;

struct PassOut {
    coords: Array3<f32>, // (n, p, 2)
    status: Array2<u8>,  // (n, p)
    err: Array2<f32>,    // (n, p)
}

/// Run LK along `frame_order` (global indices). Results are stored at the position
/// within `frame_order` (index 0 = the seed frame). Returns `None` if cancelled.
#[allow(clippy::too_many_arguments)]
fn track_pass(
    seq: &ImageSequence,
    frame_order: &[usize],
    seed: &[Point2f],
    lk: &LkParams,
    progress: &mut Option<ProgressCb>,
    total: usize,
    done0: usize,
) -> opencv::Result<Option<(PassOut, usize)>> {
    let n = frame_order.len();
    let p = seed.len();
    let mut coords = Array3::<f32>::zeros((n, p, 2));
    let mut status = Array2::<u8>::zeros((n, p));
    let mut err = Array2::<f32>::zeros((n, p));

    for (j, pt) in seed.iter().enumerate() {
        coords[[0, j, 0]] = pt.x;
        coords[[0, j, 1]] = pt.y;
        status[[0, j]] = 1;
    }

    let criteria = TermCriteria::new(
        TermCriteria_Type::EPS as i32 | TermCriteria_Type::COUNT as i32,
        lk.max_iter,
        lk.epsilon,
    )?;
    let win = Size::new(lk.win_size, lk.win_size);

    let mut prev_gray = seq.load_gray(frame_order[0])?;
    let mut prev_pts: Vector<Point2f> = seed.iter().copied().collect();
    let mut done = done0;

    for i in 1..n {
        let cur_gray = seq.load_gray(frame_order[i])?;
        let mut next_pts = Vector::<Point2f>::new();
        let mut st = Vector::<u8>::new();
        let mut er = Vector::<f32>::new();
        video::calc_optical_flow_pyr_lk(
            &prev_gray,
            &cur_gray,
            &prev_pts,
            &mut next_pts,
            &mut st,
            &mut er,
            win,
            lk.max_level,
            criteria,
            lk.flags,
            lk.min_eig_threshold,
        )?;
        for j in 0..p {
            let pt = next_pts.get(j)?;
            coords[[i, j, 0]] = pt.x;
            coords[[i, j, 1]] = pt.y;
            status[[i, j]] = st.get(j)?;
            err[[i, j]] = er.get(j)?;
        }
        prev_gray = cur_gray;
        prev_pts = next_pts; // carry forward predicted positions (failures flagged by status)
        done += 1;
        if let Some(cb) = progress.as_deref_mut() {
            if cb(done, total) {
                return Ok(None);
            }
        }
    }
    Ok(Some((PassOut { coords, status, err }, done)))
}

fn compute_fb_errors(
    coords_fw: &Array3<f32>,
    coords_bw: &Array3<f32>,
    status_fw: &Array2<u8>,
    status_bw: &Array2<u8>,
) -> (Array1<f32>, Array1<f32>) {
    let n = coords_fw.shape()[0];
    let p = coords_fw.shape()[1];
    let mut fb_mean = Array1::<f32>::from_elem(p, f32::INFINITY);
    let mut fb_max = Array1::<f32>::from_elem(p, f32::INFINITY);
    for j in 0..p {
        let mut vals: Vec<f32> = Vec::new();
        for t in 0..n {
            if status_fw[[t, j]] == 1 && status_bw[[t, j]] == 1 {
                let dx = coords_fw[[t, j, 0]] - coords_bw[[t, j, 0]];
                let dy = coords_fw[[t, j, 1]] - coords_bw[[t, j, 1]];
                vals.push((dx * dx + dy * dy).sqrt());
            }
        }
        if !vals.is_empty() {
            fb_mean[j] = vals.iter().sum::<f32>() / vals.len() as f32;
            fb_max[j] = vals.iter().copied().fold(f32::NEG_INFINITY, f32::max);
        }
    }
    (fb_mean, fb_max)
}

/// Track `seed_pts` forward then backward; returns `None` if cancelled via `progress`.
/// `coords_bw` is reindexed so `coords_bw[t]` aligns with `coords_fw[t]` (cut 0 = reference).
pub fn track(
    seq: &ImageSequence,
    reference_index: usize,
    last_index: usize,
    seed_pts: &[(f32, f32)],
    lk: &LkParams,
    mut progress: Option<ProgressCb>,
) -> opencv::Result<Option<TrackerResult>> {
    let seed: Vec<Point2f> = seed_pts.iter().map(|&(x, y)| Point2f::new(x, y)).collect();
    let n = last_index - reference_index + 1;
    let total = 2 * (n - 1);

    let forward_order: Vec<usize> = (reference_index..=last_index).collect();
    let (fw, done) = match track_pass(seq, &forward_order, &seed, lk, &mut progress, total, 0)? {
        Some(x) => x,
        None => return Ok(None),
    };

    // Backward pass: seed from forward's last-frame positions, walk last -> reference.
    let p = seed.len();
    let bw_seed: Vec<Point2f> = (0..p)
        .map(|j| Point2f::new(fw.coords[[n - 1, j, 0]], fw.coords[[n - 1, j, 1]]))
        .collect();
    let backward_order: Vec<usize> = (reference_index..=last_index).rev().collect();
    let (bw, _) = match track_pass(seq, &backward_order, &bw_seed, lk, &mut progress, total, done)? {
        Some(x) => x,
        None => return Ok(None),
    };

    // Reindex backward arrays from pass-order (last..reference) to cut-order (reference..last).
    let coords_bw = bw.coords.slice(s![..;-1, .., ..]).to_owned();
    let status_bw = bw.status.slice(s![..;-1, ..]).to_owned();
    let err_bw = bw.err.slice(s![..;-1, ..]).to_owned();

    let (fb_mean_error, fb_max_error) =
        compute_fb_errors(&fw.coords, &coords_bw, &fw.status, &status_bw);

    Ok(Some(TrackerResult {
        reference_index,
        last_index,
        coords_fw: fw.coords,
        status_fw: fw.status,
        err_fw: fw.err,
        coords_bw,
        status_bw,
        err_bw,
        fb_mean_error,
        fb_max_error,
        win_size: lk.win_size,
    }))
}
